typed/merge.go

```diff
@@ -140,81 +144,89 @@ func (w *mergingWalker) prepareDescent(pe fieldpath.PathElement, tr schema.TypeR
 func (w *mergingWalker) finishDescent(w2 *mergingWalker) {
 	// if the descent caused a realloc, ensure that we reuse the buffer
 	// for the next sibling.
-	w.errorFormatter = w2.errorFormatter.parent()
+	w.path = w2.path[:len(w2.path)-1]
 	*w.spareWalkers = append(*w.spareWalkers, w2)
 }
 
-func (w *mergingWalker) derefMap(prefix string, v *value.Value, dest **value.Map) (errs ValidationErrors) {
+func (w *mergingWalker) derefMap(prefix string, v value.Value, dest *value.Map) (errs ValidationErrors) {
 	// taking dest as input so that it can be called as a one-liner with
 	// append.
 	if v == nil {
 		return nil
 	}
-	m, err := mapValue(*v)
+	m, err := mapValue(v)
 	if err != nil {
-		return w.prefixError(prefix, err)
+		return errorf("%v: %v", prefix, err)
 	}
 	*dest = m
 	return nil
 }
 
-func (w *mergingWalker) visitListItems(t *schema.List, lhs, rhs *value.List) (errs ValidationErrors) {
-	out := &value.List{}
+func (w *mergingWalker) visitListItems(t *schema.List, lhs, rhs value.List) (errs ValidationErrors) {
+	rLen := 0
+	if rhs != nil {
+		rLen = rhs.Length()
+	}
+	lLen := 0
+	if lhs != nil {
+		lLen = lhs.Length()
+	}
+	out := make([]interface{}, 0, int(math.Max(float64(rLen), float64(lLen))))
 
 	// TODO: ordering is totally wrong.
 	// TODO: might as well make the map order work the same way.
 
 	// This is a cheap hack to at least make the output order stable.
-	rhsOrder := []fieldpath.PathElement{}
+	rhsOrder := make([]fieldpath.PathElement, 0, rLen)
 
 	// First, collect all RHS children.
-	var observedRHS fieldpath.PathElementValueMap
+	observedRHS := fieldpath.MakePathElementValueMap(rLen)
 	if rhs != nil {
-		observedRHS = fieldpath.MakePathElementValueMap(len(rhs.Items))
-		for i, child := range rhs.Items {
+		for i := 0; i < rhs.Length(); i++ {
+			child := rhs.At(i)
 			pe, err := listItemToPathElement(t, i, child)
 			if err != nil {
-				errs = append(errs, w.errorf("rhs: element %v: %v", i, err.Error())...)
+				errs = append(errs, errorf("rhs: element %v: %v", i, err.Error())...)
 				// If we can't construct the path element, we can't
 				// even report errors deeper in the schema, so bail on
 				// this element.
-				continue
+				return
```

Comment:
Why this change?

Reply:
Yeah, this one is definitely wrong, good catch

Reply:
did this cause a test to fail?

Reply:
No, probably because: 1. We don't test errors a lot, 2. It will just prevent the list of errors to be complete.